k8s/src/main/scala/io/buoyant/k8s/Watchable.scala

```diff
@@ -138,10 +160,60 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
 
     (_watch(resourceVersion), Closable.ref(close))
   }
+
+  /**
+   * Convert this Watchable into an [[Activity]]
+   * @param resourceVersion whether or not to send the initial resource
+   *                        version of the watched resource. this is a
+   *                        special case due to errors with the
+   *                        ConfigMap interpreter
+   * @param onEvent function called on each [[Watch]] event
+   * TODO: i wish this returned a [[Activity.State]] so that the caller
+   *       of this function could turn a [[Watch.Error]] into an
+   *       [[Activity.Failed]] if they wanted watch errors to fail this
+   *       [[Activity]], although none of our current code exhibits this
+   *       use-case...
+   *         - eliza, 7/18/2017
+   * @return
+   */
+  def activity[T](convert: G => T, resourceVersion: Boolean = true)
+                  (onEvent: (T, W) => T)
+  : Activity[T] =
+    Activity(Var.async[Activity.State[T]](Activity.Pending) { state =>
+      val closeRef = new AtomicReference[Closable](Closable.nop)
+      val pending = get(retryIndefinitely = true)
+        // if the initial GET failed, then the activity is a failure
+        .onFailure { e => state.update(Activity.Failed(e)) }
+        // otherwise, update the activity with the initial state, and
+        // apply the onEvent function to each successive watch event in
+        // the stream
+        .onSuccess { initial =>
+          val initialState = convert(initial)
+          state.update(Activity.Ok(initialState))
+
+          val version = if (resourceVersion) {
+            initial.metadata.flatMap(_.resourceVersion)
+          } else None
+
+          val (stream, close) = watch(None, None, version)
+
+          closeRef.set(close)
+          val _ = stream.scanLeft(initialState) { onEvent }
+            .foreach { s => state.update(Activity.Ok(s)) }
```

Comment:
Oh, gosh, that's...really bad! Glad you caught that! 

I think this bug came about because I was expecting Rust "iterator adapter" behavior here – both combinators would be called on every iteration – and I forgot Scala's iterators had different semantics. I rewrote it to use `scanLeft` because I  thought passing `onEvent` points-free was more aesthetically pleasing, but I'll definitely change it back!